xray/src/client.rs

```diff
@@ -0,0 +1,231 @@
+            self.send_encrypted(sock_dst, packet_dst, payload).await
+        } else {
+            self.send_plaintext(sock_dst, payload).await
+        }
+    }
+
+    pub async fn recv_packet(&mut self, buf: &mut [u8]) -> XRayResult<RecvType> {
+        if self.tunn.is_some() {
+            self.recv_encrypted(buf).await
+        } else {
+            self.recv_plaintext(buf).await
+        }
+    }
+
+    pub async fn tick_timers(&mut self, wg_addr: SocketAddrV4) {
+        if let Some(ref mut tunn) = &mut self.tunn {
+            match tunn.update_timers(&mut self.buf) {
+                TunnResult::Done => (),
+                TunnResult::WriteToNetwork(packet) => {
+                    if let Err(e) = self.sock.send_to(packet, wg_addr).await {
+                        println!("Update Tunn timers failed to send to socket: {:?}", e);
+                    }
+                }
+                TunnResult::Err(e) => {
+                    println!("Failed to update Tunn timers: {:?}", e);
+                }
+                unexpected_result => println!(
+                    "Update Tunn timers returned unexpected result: {:?}",
+                    unexpected_result
+                ),
+            }
+        }
+    }
+
+    async fn send_encrypted(
+        &mut self,
+        sock_dst: SocketAddrV4,
+        packet_dst: SocketAddrV4,
+        payload: &[u8],
+    ) -> XRayResult<SendType> {
+        let tunn = &mut self
+            .tunn
+            .as_mut()
+            .expect("This function should only be called on clients with a Tunn object");
+        let packet = Self::make_udp_packet(self.addr, packet_dst, payload)?;
+        let tr = tunn.encapsulate(&packet, &mut self.buf);
+        let (send_type, packet): (SendType, &[u8]) = match tr {
+            TunnResult::Done => (SendType::None, &[]),
+            TunnResult::WriteToNetwork(p) => match p[0] {
+                1 => (SendType::Handshake, p),
+                4 => (SendType::Data, p),
+                unexpected => return Err(XRayError::UnexpectedPacketType(unexpected)),
+            },
+            TunnResult::WriteToTunnelV4(p, _) => (SendType::Tunn, p),
+            unexpected => return Err(XRayError::from(unexpected)),
+        };
+        if !matches!(send_type, SendType::None) {
+            self.sock.send_to(packet, sock_dst).await?;
+        }
+        Ok(send_type)
+    }
+
+    async fn send_plaintext(&mut self, dst: SocketAddrV4, payload: &[u8]) -> XRayResult<SendType> {
+        self.sock.send_to(payload, dst).await?;
+        Ok(SendType::Plaintext)
+    }
+
+    async fn recv_encrypted(&mut self, buf: &mut [u8]) -> XRayResult<RecvType> {
+        let (mut bytes_read, from) = self.sock.recv_from(&mut self.buf).await?;
+        let from = match from {
+            SocketAddr::V4(addr) => addr,
+            SocketAddr::V6(_) => return Err(XRayError::Ipv6),
+        };
+        let mut ret = match self.buf[0] {
+            2 => RecvType::Handshake,
```

Comment:
So you always expect this side to be initiator and then a receiver? It look a bit confusing to have 1 and 2 map to handshake instead of HandshakeInitiation / HandshakeResponse

Reply:
The idea was to be the initiator of handshakes but it should of course work with being the receiver as well. I changed the `SendType` and `RecvType` to have `HandshakeInitiation` and `HandshakeResponse` instead of just `Handshake`. I haven't actively tried if it works to be on the responder side of a handshake but I think it should work